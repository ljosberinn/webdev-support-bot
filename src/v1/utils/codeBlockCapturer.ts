const BACKTICKS = '```';

export type CodeBlockData = {
  code: string;
  language: string;
};

export function createCodeBlockCapturer(langs: string[] = []) {
  const langAlts = langs.join('|');
  return function* (str: string): Generator<CodeBlockData> {
    const langRegex = new RegExp(
      String.raw`${BACKTICKS}(?<language>${langAlts})\n(?<code>[\s\S]+?)\n${BACKTICKS}`,
      'gi'
    );
    const matches = str.matchAll(langRegex);
    for (const {
      groups: { language, code },
    } of matches) {
      yield {
        language: language || '',
        code: code || '',
      };
    }
  };
}
