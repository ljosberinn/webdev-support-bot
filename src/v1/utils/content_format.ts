import { Message } from 'discord.js';

const linebreakPattern = /\n/gim;

export const generateCleanContent = (msg: Message) =>
  msg.cleanContent.replace(linebreakPattern, ' ').toLowerCase();

export const stripMarkdownQuote = (msg: string) => msg.replace(/^> .+$/gm, '');
