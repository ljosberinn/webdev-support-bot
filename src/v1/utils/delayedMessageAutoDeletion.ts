import { Message } from 'discord.js';

import { missingRightsDeletion } from './errors';

const THIRTY_SECONDS_IN_MS = 30 * 1000;

export const delayedMessageAutoDeletion = (
  msg: Message,
  timeout = THIRTY_SECONDS_IN_MS
) => {
  // required so tests on CI dont crash
  if (msg) {
    setTimeout(async () => {
      try {
        await msg.delete();
      } catch (error) {
        // eslint-disable-next-line no-console
        console.warn("Couldn't delete message", error);

        try {
          await msg.edit(missingRightsDeletion);
        } catch {
          // eslint-disable-next-line no-console
          console.info(
            "Couldn't edit message after trying to delete, probably removed by someone else."
          );
        }
      }
    }, timeout);
  }

  return;
};
