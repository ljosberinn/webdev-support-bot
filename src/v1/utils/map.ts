/**
 * same as Array#map but works on all iterables, returns an iterable
 *
 * @export
 * @template T input iterable
 * @template U output iterable
 * @param {(item: T, index: number, iter: Iterable<T>) => U} fn
 * @param {Iterable<T>} iter
 */
export function* mapʹ<T, U>(
  fn: (item: T, index: number, iter: Iterable<T>) => U,
  iter: Iterable<T>
) {
  let i = 0;
  for (const item of iter) {
    yield fn(item, i++, iter);
  }
}

export function map<T, U>(
  fn: (item: T, index: number, iter: Iterable<T>) => U
) {
  return (iter: Iterable<T>) => mapʹ(fn, iter);
}
