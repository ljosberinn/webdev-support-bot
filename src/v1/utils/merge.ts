export function* merge<T>(...iterables: Iterable<T>[]) {
  for (const iterable of iterables) {
    yield* iterable;
  }
}
