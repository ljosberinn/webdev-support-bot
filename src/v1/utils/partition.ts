export function partitionʹ<T, U>(
  predicate: (item: T, index: number, iter: Iterable<T>) => U,
  iter: Iterable<T>
) {
  const success = [];
  const failed = [];
  let i = 0;
  for (const item of iter) {
    const arr = predicate(item, i++, iter) ? success : failed;
    arr.push(item);
  }
  return [success, failed];
}

export function partition<T, U>(
  predicate: (item: T, index: number, iter: Iterable<T>) => U
) {
  return function (iter: Iterable<T>) {
    return partitionʹ(predicate, iter);
  };
}
