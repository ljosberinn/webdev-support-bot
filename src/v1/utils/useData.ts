import * as Sentry from '@sentry/node';
import fetch, {
  HeaderInit,
  RequestInfo,
  RequestInit,
  Response,
} from 'node-fetch';

import {
  API_CACHE_ENTRIES_LIMIT,
  API_CACHE_EXPIRATION_IN_SECONDS,
  API_CACHE_REVALIDATION_WINDOW_IN_SECONDS,
} from '../env';
import { Cache } from './Cache';

const apiCache = new Cache({
  checkperiod: Number.parseInt(API_CACHE_REVALIDATION_WINDOW_IN_SECONDS, 10),
  maxKeys: Number.parseInt(API_CACHE_ENTRIES_LIMIT, 10),
  stdTTL: Number.parseInt(API_CACHE_EXPIRATION_IN_SECONDS, 10),
});

type ResponseTypes = 'json' | 'text';

type UnknownData<T> =
  | {
      error: true;
      json: null;
      text: null;
    }
  | {
      error: false;
      text: null;
      json: T;
    }
  | {
      error: false;
      json: null;
      text: string;
    };

type ResponseMapper<Data> = (response: Response) => Promise<Data>;
type FetchWithFormat<Format> = (
  url: RequestInfo,
  init?: RequestInit
) => Promise<Format>;

/**
 * Given a cache key and a function that maps the fetch response to a arbitrary
 * data structure, return a function when invoked, lower cases the cache key and
 * returns the cached formatted response when found.
 *
 * If the cache entry is _not_ found, delegate the call to an actual fetch implementation
 * to make the HTTP request. This response is then formatted with the function
 * above and saved into the cache. HTTP requests times/durations are also logged.
 */
const doFetch: <TParsedResponse>(
  cacheKey: string,
  mapper: ResponseMapper<TParsedResponse>
) => FetchWithFormat<TParsedResponse> = <TParsedResponse>(cacheKey, mapper) => {
  const casedCacheKey = cacheKey.toLowerCase();
  const cachedResponse = apiCache.get<TParsedResponse>(casedCacheKey);

  if (cachedResponse) {
    return () => cachedResponse;
  }

  return async (url, fetchOptions) => {
    Sentry.addBreadcrumb({
      category: 'query',
      data: typeof url === 'string' ? { url } : undefined,
      level: Sentry.Severity.Info,
      timestamp: Date.now(),
    });

    const timeLabel = `Time took for url=${encodeURIComponent(url.toString())}`;
    // eslint-disable-next-line no-console
    console.time(timeLabel);
    const response = await fetch(url, fetchOptions);
    // eslint-disable-next-line no-console
    console.timeEnd(timeLabel);
    const formattedResponse = await mapper(response);
    apiCache.set(casedCacheKey, formattedResponse);
    return formattedResponse;
  };
};

const responseMapper: <T>(
  type: ResponseTypes
) => (response: Response) => Promise<UnknownData<T>> = (
  type: string
) => async response => {
  if (!response.ok) {
    return {
      error: true,
      json: null,
      text: null,
    };
  }

  if (type === 'json') {
    const json = await response.json();
    return {
      error: false,
      json,
      text: null,
    };
  }

  if (type === 'text') {
    const text = await response.text();

    return {
      error: false,
      json: null,
      text,
    };
  }
};

const useData = <T>(
  url: string,
  type: ResponseTypes = 'json',
  headers: HeaderInit = {}
): Promise<UnknownData<T>> => {
  return doFetch(url, responseMapper<T>(type))(url, { headers });
};

export default useData;
