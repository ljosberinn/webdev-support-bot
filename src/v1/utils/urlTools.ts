import { Message } from 'discord.js';
import { HeadersInit } from 'node-fetch';

import { delayedMessageAutoDeletion } from './delayedMessageAutoDeletion';
import { Provider } from './discordTools';
import { noResults, invalidResponse } from './errors';
import useData from './useData';

const SEARCH_TERM = '%SEARCH%';
const TERM = '%TERM%';

type ProviderMap = {
  [key in Provider]: {
    search: string;
    color: number;
    createTitle: (term: string) => string;
    icon: string;
    help: string;
    direct?: string;
    getExtendedInfoUrl?: (pkg: string) => string;
  };
};

export const providers: ProviderMap = {
  bundlephobia: {
    color: 0xffffff,
    createTitle: (searchTerm: string) =>
      `Bundlephobia results for *${searchTerm}*`,
    direct: `https://bundlephobia.com/result?p=${TERM}`,
    getExtendedInfoUrl: (pkg: string) =>
      `https://bundlephobia.com/api/size?package=${pkg}&record=true`,
    help: '!bundlephobia @chakra-ui/core',
    icon: 'https://bundlephobia.com/android-chrome-192x192.png',
    search: `https://api.npms.io/v2/search/suggestions?q=${SEARCH_TERM}`,
  },
  caniuse: {
    color: 0xdb5600,
    createTitle: (searchTerm: string) => `CanIUse results for *${searchTerm}*`,
    direct: `https://caniuse.com/#feat=${TERM}`,
    getExtendedInfoUrl: (text: string) =>
      `https://caniuse.com/process/get_feat_data.php?type=support-data&feat=${text}`,
    help: '!caniuse IntersectionObserver',
    icon: 'https://caniuse.com/img/favicon-128.png',
    search: `https://caniuse.com/process/query.php?search=${SEARCH_TERM}`,
  },
  composer: {
    color: 0xf28d1a,
    createTitle: (searchTerm: string) => `Packagist results for ${searchTerm}`,
    direct: `https://packagist.org/packages/${TERM}`,
    getExtendedInfoUrl: (pkg: string) =>
      `https://packagist.org/packages/${pkg}.json`,
    help: '!composer sentry',
    icon: 'https://packagist.org/bundles/packagistweb/img/logo-small.png',
    search: `https://packagist.org/search.json?q=${SEARCH_TERM}`,
  },
  github: {
    color: 0x24292e,
    createTitle: (searchTerm: string) => `GitHub results for *${searchTerm}*`,
    direct: `https://github.com/${TERM}`,
    help: '!github react',
    icon:
      'https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png',
    search: `https://api.github.com/search/repositories?q=${SEARCH_TERM}`,
  },
  mdn: {
    color: 0x83d0f2,
    createTitle: (searchTerm: string) => `MDN results for *${searchTerm}*`,
    direct: `https://developer.mozilla.org${TERM}`,
    help: '!mdn localStorage',
    icon: 'https://avatars0.githubusercontent.com/u/7565578',
    search: `https://developer.mozilla.org/api/v1/search?q=${SEARCH_TERM}&locale=en-US`,
  },
  npm: {
    color: 0xfb3e44,
    createTitle: (searchTerm: string) => `NPM results for *${searchTerm}*`,
    help: '!npm react',
    icon: 'https://avatars0.githubusercontent.com/u/6078720',
    search: `https://www.npmjs.com/search/suggestions?q=${SEARCH_TERM}`,
  },
  php: {
    color: 0x8892bf,
    createTitle: (searchTerm: string) => `PHP.net results for *${searchTerm}*`,
    direct: `https://www.php.net/${TERM}`,
    help: '!php echo',
    icon: 'https://www.php.net/images/logos/php-logo.svg',
    search: `https://www.php.net/${SEARCH_TERM}`,
  },
};

export const HELP_KEYWORD = '--help';
export const FORMATTING_KEYWORD = '!formatting';
export const FORMATTING_KEYWORD_ALT = '!format';
export const CODE_KEYWORD = '!code';
export const VSCODE_KEYWORD = '!vscode';
export const JOB_POSTING_KEYWORD = '!post';
export const JQUERY_KEYWORD = '!jquery';
export const POINTS_KEYWORD = '!points';
export const LEADERBOARD_KEYWORD = '!leaderboard';
export const DECAY_KEYWORD = '!decay';
export const MODULE_KEYWORD = '!modules';
export const LOCKFILE_KEYWORD = '!lockfile';
export const FLEXBOX_KEYWORD = '!flexbox';

/**
 * dynamic regExp matching all possible Object.keys(providers) as keyword
 */
export const KEYWORD_REGEXP = new RegExp(
  '^!(%PLACEHOLDER%)\\s+'.replace(
    '%PLACEHOLDER%',
    Object.keys(providers)
      .reduce((carry, keyword) => [...carry, keyword], [])
      .join('|')
  ),
  'i'
);

export const getSearchUrl = (provider: Provider, search: string) => {
  if (providers[provider]) {
    return providers[provider].search.replace(SEARCH_TERM, encodeURI(search));
  }

  throw new Error(`provider not implemeted: ${provider}`);
};

export const buildDirectUrl = (provider: Provider, href: string) => {
  if (providers[provider]) {
    return providers[provider].direct.replace(TERM, href);
  }

  throw new Error(`provider not implemeted: ${provider}`);
};

export const getExtendedInfoUrl = (provider: Provider, term: string) => {
  if (providers[provider] && providers[provider].getExtendedInfoUrl) {
    return providers[provider].getExtendedInfoUrl(term);
  }

  throw new Error(
    `provider or provider.getExtendedInfoUrl not implemented at provider: ${provider}`
  );
};

interface GetDataParams {
  msg: Message;
  provider: Provider;
  searchTerm: string;
  sanitizeData?: (data: any) => Partial<any>;
  isInvalidData: (data: any) => boolean;
  headers?: HeadersInit;
}

export const getData = async <T>({
  msg,
  provider,
  searchTerm,
  sanitizeData,
  isInvalidData,
  headers,
}: GetDataParams): Promise<Partial<T>> => {
  const searchUrl = getSearchUrl(provider, searchTerm);
  const { error, json: data } = await useData<T>(searchUrl, 'json', headers);

  if (error) {
    await msg.reply(invalidResponse);
    return;
  }

  const sanitizedData = sanitizeData ? sanitizeData(data) : data;

  if (isInvalidData(sanitizedData)) {
    const sentMessage = await msg.reply(noResults(searchTerm));

    delayedMessageAutoDeletion(sentMessage);
    return;
  }

  return sanitizedData;
};
