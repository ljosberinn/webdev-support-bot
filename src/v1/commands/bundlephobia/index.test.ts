import { getChosenResult } from '../../utils/discordTools';
import * as errors from '../../utils/errors';
import { getData } from '../../utils/urlTools';
import useData from '../../utils/useData';
import {
  response,
  detailResponse,
  dateFn,
  similar,
  previousVersion,
} from './__fixtures__/response';

import { buildBundlephobiaQueryHandler } from '.';

describe('buildBundlephobiaQueryHandler', () => {
  const editMock = jest.fn();
  const replyCallback = jest.fn();
  const sendMock = jest.fn(() =>
    Promise.resolve({
      edit: editMock,
      reply: replyCallback,
    })
  );

  const replyMock = jest.fn();
  const msg: any = {
    channel: { send: sendMock },
    delete: replyMock,
    reply: replyMock,
  };

  const fetch: jest.MockedFunction<typeof getData> = jest.fn();
  const fetchDetail: jest.MockedFunction<typeof useData> = jest.fn();
  const choose: jest.MockedFunction<typeof getChosenResult> = jest.fn();

  beforeEach(() => {
    fetch.mockResolvedValue(response);
  });

  afterEach(() => jest.clearAllMocks());

  test('fails when API request responds with error', async () => {
    fetch.mockResolvedValueOnce(null);
    const handler = buildBundlephobiaQueryHandler(fetch, fetchDetail, choose);
    await handler(msg, 'moment.js');

    expect(choose).not.toBeCalled();
    expect(msg.channel.send).not.toBeCalled();
  });

  test('works when API request responds, but fails if the reply dies', async () => {
    choose.mockResolvedValueOnce(null);
    const handler = buildBundlephobiaQueryHandler(fetch, fetchDetail, choose);
    await handler(msg, 'moment.js');
    expect(fetchDetail).not.toBeCalled();
  });

  test('works when reply response, but dies on detail', async () => {
    choose.mockResolvedValueOnce({
      description: 'Parse, validate, manipulate, and display dates',
      name: 'moment',
      url: 'http://big.thick.dates',
    });

    fetchDetail.mockResolvedValueOnce({
      error: true,
      json: null,
      text: null,
    });

    const handler = buildBundlephobiaQueryHandler(fetch, fetchDetail, choose);
    await handler(msg, 'moment.js');

    expect(msg.reply).toBeCalledWith(errors.invalidResponse);
  });

  test('works when API request responds and details responds', async () => {
    choose.mockResolvedValueOnce({
      description: 'Parse, validate, manipulate, and display dates',
      name: 'moment',
      url: 'http://big.thick.dates',
    });

    // Mock the implementation since there's no way in hell I am going to
    // extract each call into a service function.
    fetchDetail.mockImplementation(
      (url: string) =>
        new Promise(resolve => {
          if (url.includes('similar-packages')) {
            resolve({
              error: false,
              json: similar,
              text: null,
            });
          }

          if (url.includes('date-fn')) {
            resolve({
              error: false,
              json: dateFn,
              text: null,
            });
          }

          if (url.includes('package-history')) {
            resolve({
              error: false,
              json: previousVersion,
              text: null,
            });
          }

          resolve({
            error: false,
            json: detailResponse,
            text: null,
          });
        })
    );

    const handler = buildBundlephobiaQueryHandler(fetch, fetchDetail, choose);
    await handler(msg, 'moment.js');

    expect(editMock.mock.calls[0]).toBeDefined();
    expect(editMock.mock.calls[0][0]).toMatchSnapshot();
  });
});
