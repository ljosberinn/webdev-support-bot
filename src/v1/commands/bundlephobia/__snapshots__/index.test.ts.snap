// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildBundlephobiaQueryHandler works when API request responds and details responds 1`] = `
Object {
  "embed": Object {
    "author": null,
    "color": 16777215,
    "description": "Parse, validate, manipulate, and display dates",
    "fields": Array [
      Object {
        "inline": true,
        "name": "size (latest)",
        "value": "69.90kb
(+0.69% to last)",
      },
      Object {
        "inline": true,
        "name": "size (previous)",
        "value": "69.42kb",
      },
      Object {
        "inline": true,
        "name": "dependencies",
        "value": "0",
      },
      Object {
        "inline": true,
        "name": "​",
        "value": "​",
      },
      Object {
        "inline": true,
        "name": "treeshakable?",
        "value": ":x:",
      },
      Object {
        "inline": true,
        "name": "side effects?",
        "value": ":exclamation: yes",
      },
      Object {
        "inline": true,
        "name": "download on edge",
        "value": "2.33s",
      },
      Object {
        "inline": true,
        "name": "download on 3g",
        "value": "1.40s",
      },
      Object {
        "inline": true,
        "name": "​",
        "value": "​",
      },
      Object {
        "inline": false,
        "name": "Similar packages",
        "value": "labelled as _General purpose date-time utilities_",
      },
      Object {
        "inline": true,
        "name": "date-fns",
        "value": "17.88kb",
      },
    ],
    "footer": Object {
      "iconURL": "https://bundlephobia.com/android-chrome-192x192.png",
      "text": "Sizes are always gzipped.",
    },
    "title": "moment",
    "url": "https://bundlephobia.com/result?p=moment",
  },
}
`;
